odl/test/space/fspace_test.py

```diff
@@ -110,683 +49,1111 @@ def _meshgrid(domain, shape):
+    assert all_almost_equal(result_point2, true_result_point)
+
+    out_points = np.empty(values_points_shape, dtype=float)
+    out_mesh = np.empty(values_mesh_shape, dtype=float)
+    out_point1 = np.empty(value_point_shape, dtype=float)
+    out_point2 = np.empty(value_point_shape, dtype=float)
+    func_elem(points, out=out_points)
+    func_elem(mesh, out=out_mesh)
+    func_elem(point1, out=out_point1)
+    func_elem(point2, out=out_point2)
+    assert all_almost_equal(out_points, true_result_points)
+    assert all_almost_equal(out_mesh, true_result_mesh)
+    assert all_almost_equal(out_point1, true_result_point)
+    assert all_almost_equal(out_point2, true_result_point)
+
+
+def test_fspace_elem_equality():
+    """Test equality check of fspace elements."""
+    intv = odl.IntervalProd(0, 1)
+    fspace = FunctionSpace(intv)
+
+    f_novec = fspace.element(func_nd_oop, vectorized=False)
+
+    f_vec_oop = fspace.element(func_nd_oop, vectorized=True)
+    f_vec_oop_2 = fspace.element(func_nd_oop, vectorized=True)
+
+    f_vec_ip = fspace.element(func_nd_ip, vectorized=True)
+    f_vec_ip_2 = fspace.element(func_nd_ip, vectorized=True)
+
+    f_vec_dual = fspace.element(func_nd_dual, vectorized=True)
+    f_vec_dual_2 = fspace.element(func_nd_dual, vectorized=True)
+
+    assert f_novec == f_novec
+    assert f_novec != f_vec_oop
+    assert f_novec != f_vec_ip
+    assert f_novec != f_vec_dual
+
+    assert f_vec_oop == f_vec_oop
+    assert f_vec_oop == f_vec_oop_2
+    assert f_vec_oop != f_vec_ip
+    assert f_vec_oop != f_vec_dual
+
+    assert f_vec_ip == f_vec_ip
+    assert f_vec_ip == f_vec_ip_2
+    assert f_vec_ip != f_vec_dual
+
+    assert f_vec_dual == f_vec_dual
+    assert f_vec_dual == f_vec_dual_2
+
+    fspace_tens = FunctionSpace(intv, out_dtype=(float, (2, 3)))
+
+    f_tens_oop = fspace_tens.element(func_tens_oop)
+    f_tens_oop2 = fspace_tens.element(func_tens_oop)
+
+    f_tens_ip = fspace_tens.element(func_tens_ip)
+    f_tens_ip2 = fspace_tens.element(func_tens_ip)
+
+    f_tens_seq = fspace_tens.element(func_tens_oop_seq)
+    f_tens_seq2 = fspace_tens.element(func_tens_oop_seq)
+
+    assert f_tens_oop == f_tens_oop
+    assert f_tens_oop == f_tens_oop2
+    assert f_tens_oop != f_tens_ip
+    assert f_tens_oop != f_tens_seq
+
+    assert f_tens_ip == f_tens_ip
+    assert f_tens_ip == f_tens_ip2
+    assert f_tens_ip != f_tens_seq
+
+    # Sequences are wrapped, will compare to not equal
+    assert f_tens_seq == f_tens_seq
+    assert f_tens_seq != f_tens_seq2
+
+
+def test_fspace_elem_assign(out_shape):
+    """Check assignment of fspace elements."""
+    fspace = FunctionSpace(odl.IntervalProd(0, 1),
+                           out_dtype=(float, out_shape))
+
+    ndim = len(out_shape)
+    if ndim == 0:
+        f_oop = fspace.element(func_nd_oop)
+        f_ip = fspace.element(func_nd_ip)
+        f_dual = fspace.element(func_nd_dual)
+    elif ndim == 1:
```

Comment:
Second thought, defining a tensor-valued function with non-square and non-trivial shape requires at least `(2, 3)`, which makes it quite a hassle to do for all cases relevant in these tests. So I took shape `(2,)` for the comprehensive tests and added just a few cases for shape `(2, 3)`.